Bootstrap themes

joankaradimov opened the pull request:
I've hacked together a simple implementation of #2334. It has some visual issues, but it's usable overall.

If this is still on the roadmap I could spend some time improving the implementation.

arshaw commented:
@joankaradimov this is really awesome! Can I offer some suggestions on how to make the code better, so that I can include it in a release?

All over the place there are statements like this:
```
FC.ThemeClasses.stateDown(tm)
```

It'd be great if we could rely on a method more closely related to the current object so that we don't need to query for the `tm` value each time. So:

```
// from within the calendar class
this.theme.stateDown();

// from within the view class
this.calendar.theme.stateDown();

// from within a Grid class
this.view.calendar.theme.stateDown()
```

So, you'd need to create a `theme` object that becomes a property of the calendar. The Calendar would be responsible for querying for the `theme` option ahead of time.

Make sense?

joankaradimov commented:
> Make sense?

It makes perfect sense. I hadn't realized that a calendar instance is always accessible from the scope of all objects that use themes.

joankaradimov commented:
Done. Thanks for the feedback. Tell me if/when you want me to squash.

One thing I was thinking about - I could iterate the `themes` hash and dynamically generate the functions for html classes in `Theme.js`. It would be shorter but less explicit. Any preferences on that one?

Any other feedback?

arshaw commented:
hi @joankaradimov , sorry for the delay. I prefer `this.calendar.theme.getClass('widgetHeader')` instead of `this.calendar.theme.widgetHeader()` . If you don't have time to make the changes, I can.

joankaradimov commented:
No problem. It's a quick change. I'll do it.

joankaradimov commented:
Done. The last test failure is the same as the other recent PRs and shouldn't happen when the tests are ran locally.

criztovyl commented on src/common/Theme.js:
```diff
@@ -0,0 +1,58 @@
+		} else {
+			this.theme = 'builtin';
+		}
+	},
```
I don't understand what the `if...else if...else` does, why hardcode possible themes there?

joankaradimov replied:
The intent was to keep compatibility with the old `true`/`false`/`undefined` parameter. But I see why it's not obvious what this code does. I've updated the pull request.

arshaw commented:
Hi @joankaradimov , I was about to merge this into the upcoming release myself (was going to resolve the inevitable merge conflicts too), but decided to hold off because there are some QA issues:

- relies on jQuery UI button icons. i would rather not need people to include jqueryui+bootstrap theme CSS on the same page.
- in the demo it would really really nice to allow the user to select a theme manually, rather than relying on the random button
- harsh dark lines in agenda view: ![screen shot 2017-04-26 at 3 55 41 pm](https://cloud.githubusercontent.com/assets/56520/25454092/50ce8414-2a99-11e7-8b23-43464441be98.png)
- popover is broken: ![screen shot 2017-04-26 at 4 01 58 pm](https://cloud.githubusercontent.com/assets/56520/25454231/bb21114c-2a99-11e7-80d7-fc1e8b6bb347.png)

arshaw commented on src/common/Theme.js:
```diff
@@ -0,0 +1,58 @@
+var themes = {
+    builtin: {
```
mixing tabs and spaces

joankaradimov replied:
I'll fix and force push this particular piece code.

joankaradimov commented:
> was going to resolve the inevitable merge conflicts too

I can do the rebase. I am not doing it now, though. I'll need some time to get it right.

> relies on jQuery UI button icons. i would rather not need people to include jqueryui+bootstrap theme CSS on the same page.

Using Glyphicons would be the easiest solution here. But I'm not fond of it as some people disable Glyphicons and use alternative solutions (me being one of those people). Let me think about it.

> in the demo it would really really nice to allow the user to select a theme manually, rather than relying on the random button

Done.

> harsh dark lines in agenda view: 

Done.

> popover is broken

I am yet to look at this one.

arshaw commented:
sounds great @joankaradimov, let me know if you have any questions.

joankaradimov commented:
I squashed, rebased, fixed some of the issues, and force pushed. Maybe a new review is in order.

What remains to be done is to figure out what to do with the button icons.

joankaradimov commented on src/Calendar.render.js:
```diff
@@ -51,6 +51,7 @@ Calendar.mixin({
 		this.optionsModel.watch('applyingThemeClasses', [ '?theme' ], function(opts) {
 			el.toggleClass('ui-widget', opts.theme);
 			el.toggleClass('fc-unthemed', !opts.theme);
+			_this.theme.setTheme(opts.theme);
```
I am not sure this is the correct way to do this now. Is there a cleverer solution here?

joankaradimov commented:
I've updated the pull request.

The button icons are now part of the theme. If you pick the builtin theme or the jQuery UI theme you can manipulate the icons as before (via the [`buttonIcons`](https://fullcalendar.io/docs/display/buttonIcons/) and [`themeButtonIcons`](https://fullcalendar.io/docs/display/themeButtonIcons/) settings accordingly). The `icon` and `themeIcon` properties of [`customButtons`](https://fullcalendar.io/docs/display/customButtons/) should also work as before. Setting these will override the default icons.

Currently it is not possible to manipulate the icons of the new bootstrap theme. It should be fairly straightforward to expose an interface that allows the manipulation of the icons (maybe extending existing themes or creating new ones).

However, such a mechanism would be redundant with the backwards compatible one. So I've left it out for now. A suggestion on how to go with this would be welcome. If it's up to me I would deprecate the old way of setting icons and keep it for a while along a new way.

I've also created two separate bootstrap themes - one that uses glyphicons and one that does not.

arshaw commented:
hi @joankaradimov, I did some additional work on top your branch, the first part of which can be seen [in these commits](https://github.com/fullcalendar/fullcalendar/commits/8edbe9c252fe0bd625ba57484c701898dff70fde).

- Made each theme system more plugin-esque
- rnamed the theme `bootstrap3`
- new setting called `bootstrapGlyphicons` which can customize individual icons
- if you don't want to use glyphicons, set `bootstrapGlyphicons` to `false`. I think this achieves what you want, though it does not default the icons to the built-in theme's icons
- rename `built-in` theme to `standard`

I then merged all this into [the master branch](https://github.com/fullcalendar/fullcalendar/commits/master), which contains a big refactor of other misc stuff.

After that, the biggest thing I did was redo the demo page. It combines jQuery UI and Bootstrap into one demo file. Plz take a look (`demos/themes.html`).

This PR is now closed because it has been merged into master, but it still needs to be released. There's a bit more refactoring of other subsystems before the next release. If you want to make further changes, please base PRs against `master`.

joankaradimov commented:
Thanks!

> If you want to make further changes, please base PRs against master.

The only thing I'd like to suggest is that the color of the "today" cell is tweaked in the Bootstrap theme. I made this PR for it: https://github.com/fullcalendar/fullcalendar/pull/3745.

Other than that - it looks like these two issues can be closed now: 
 - https://github.com/fullcalendar/fullcalendar/issues/2334
 - https://github.com/fullcalendar/fullcalendar/issues/1436

Also the author of the Material Design Theme in https://github.com/fullcalendar/fullcalendar/issues/2836 might find it much easier to integrate his changes into master now.

And finally - I opened https://github.com/fullcalendar/fullcalendar/issues/3564 a while ago. I'll try to figure out what the separation of `common.css` into several files means for the issue. If there's only 5-6 KB to be gained, it should not be considered at all. But if more CSS can be split in a similar fashion, I'll be interested in making the pull requests.

arshaw commented:
@joankaradimov thanks for all the hard work coding as well as helping me keep track of these issues.

I've merged the "today" PR and commented on all the other ones you mentioned.

joankaradimov commented:
@arshaw Thanks for the cooperation and the guidance. And for the great project, of course!

I'm looking forward to the next release.

arshaw commented:
Released in v3.5.0 ([blog post](https://fullcalendar.io/blog/2017/08/bootstrap-support-and-future-developments/))

thanks @joankaradimov 

If you're curious, the delay in releasing after this was implemented was due to me trying to do continued refactoring after the merge. However, the refactoring became much more extensive than planned, so I decided to release the code at the time of the Bootstrap 3 merge, which was stable. I'm continuing to work on the refactor. More info on the blog.

joankaradimov commented:
> If you're curious, the delay in releasing after this was implemented was due to me trying to do continued refactoring after the merge

Yes, I've followed the commits. I'd love to see this progress so we can see more features and performance improvements in the future.

And considering the blog post - Bootstrap 4 support sounds great. I'd love to implement it once we have a final version (and hopefully support for the Bootswatch themes).
